// @superblocksteam/util/dist/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.NotFoundError = exports.FileAccessError = exports.CUSTOM_COMPONENTS_PATH = exports.ROOT_CONFIG_RELATIVE_PATH = exports.RESOURCE_CONFIG_PATH = exports.TOKEN_CONFIG_PATH = exports.SUPERBLOCKS_HOME_FOLDER_NAME = void 0;
exports.SUPERBLOCKS_HOME_FOLDER_NAME = ".superblocks";
exports.TOKEN_CONFIG_PATH = ".superblocks/auth.json";
exports.RESOURCE_CONFIG_PATH = ".superblocks/superblocks.json";
/* It's a relative path from resource's directory to a project's root
├── apps
│   ├── <app_name>
│   │   ├── apis
│   │   │   ├── <api_name1>
...
│   │   │   ├── <api_nameN>
│   │   ├── application.yaml
│   │   └── page.yaml
├── backends
│   └── <scheduled job|workflow_name>
│       └── api.yaml
*/
exports.ROOT_CONFIG_RELATIVE_PATH = "../..";
exports.CUSTOM_COMPONENTS_PATH = "components";
class FileAccessError extends Error {
    constructor(message) {
        super(message);
        this.name = "FileAccessError";
    }
}
exports.FileAccessError = FileAccessError;
class NotFoundError extends Error {
    constructor(message) {
        super(message);
        this.name = "FileNotFoundError";
    }
}
exports.NotFoundError = NotFoundError;
